curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java

```diff
@@ -228,15 +234,18 @@ protected void     internalRegisterService(ServiceInstance<T> service) throws Ex
     public void     unregisterService(ServiceInstance<T> service) throws Exception
     {
         String          path = pathForInstance(service.getName(), service.getId());
-        try
-        {
-            client.delete().forPath(path);
-        }
-        catch ( KeeperException.NoNodeException ignore )
-        {
-            // ignore
+        synchronized(service) {
```

Comment:
Imagine the following sequence:

T1: reconnect()
T2: unregister(X)
T1: for each service in cache:
T1:   register(X)
T2: services.remove(X)
T2: client.delete(X)
T1: client.create(X)